### PR TITLE
thrift_proxy: configurable decoder filters

### PR DESCRIPTION
**Pull request by zuercher**
Modifies the thrift proxy to allow configurable decoder filters with
the ability for any filter to pause/resume request decoding (e.g. to
response to an external event) or to short-circuit request handling
with a locally-generated response. (The primary motivation for these
features are a rate-limiting thrift filter that will follow separately.)

*Risk Level*: low
*Testing*: unit tests
*Docs Changes*: inline
*Release Notes*: updated

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

**Comment by zuercher** [comment]
@brirams @rg1 @fishcakez if you'd like to take a look

File: source/extensions/filters/network/thrift_proxy/conn_manager.h [hunk 1]
```diff
@@ -87,8 +89,8 @@ class ConnectionManager : public Network::ReadFilter,
 
     // ProtocolConverter
     FilterStatus messageBegin(MessageMetadataSharedPtr metadata) override;
-    FilterStatus fieldBegin(absl::string_view name, FieldType field_type,
-                            int16_t field_id) override;
+    FilterStatus fieldBegin(absl::string_view name, FieldType& field_type,
+                            int16_t& field_id) override;
```
Comment (brian-pane) [inline source/extensions/filters/network/thrift_proxy/conn_manager.h]: What's the scenario where an implementation of `fieldBegin` will need to modify the `field_type` or `field_id`?

Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/conn_manager.h]: The use case is some complex filter that's manipulating thrift structs. I could make this simpler by reverting this and not allowing thrift filters to manipulate the request beyond the message metadata.

Comment (brian-pane) [reply source/extensions/filters/network/thrift_proxy/conn_manager.h]: Got it, thanks. I guess making the Thrift structs mutable by filters is consistent with how HTTP filters can mutate message bodies, so changing these parameters to references seems reasonable.

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc [hunk 2]
```diff
@@ -243,45 +246,137 @@ FilterStatus ConnectionManager::ResponseDecoder::transportEnd() {
...
 FilterStatus ConnectionManager::ActiveRpc::transportEnd() {
   ASSERT(metadata_ != nullptr);
-  ASSERT(metadata_->hasMessageType());
+
+  FilterStatus status;
+  if (upgrade_handler_) {
+    status = upgrade_handler_->transportEnd();
+
+    if (metadata_->isProtocolUpgradeMessage()) {
+      ENVOY_CONN_LOG(error, "thrift: sending protocol upgrade response",
```
Comment (bwangelme) [inline source/extensions/filters/network/thrift_proxy/conn_manager.cc:318]: @zuercher 

Hello, I have a question about this.

Why did you change the level of this log sentence (`thrift: sending protocol upgrade response`) from debug to Error

This seems to have returned a Twitter upgrade response normally, not an error.
